crates/proxy-server/src/mod_proxy.rs

```diff
@@ -0,0 +1,244 @@
+                tracing::error!("accept loop returned with error: {err:#}");
+            }
+        })?;
+
+        Ok(())
+    }
+
+    async fn accept_loop(
+        listener: TcpListener,
+        params: Arc<Self>,
+        tls_acceptor: Option<TlsAcceptor>,
+    ) -> anyhow::Result<()> {
+        let local_address = listener
+            .local_addr()
+            .context("failed to get local address")?;
+
+        loop {
+            let (socket, peer_address) = listener.accept().await.context("accept failed")?;
+
+            let params = params.clone();
+            let tls_acceptor = tls_acceptor.clone();
+
+            tokio::spawn(async move {
+                let result: anyhow::Result<()> = async {
+                    if let Some(acceptor) = tls_acceptor {
+                        let tls_stream = acceptor.accept(socket).await.with_context(|| {
+                            format!("failed TLS handshake from {peer_address:?}")
+                        })?;
+                        Self::handle_client(tls_stream, peer_address, local_address, &params).await
+                    } else {
+                        Self::handle_client(socket, peer_address, local_address, &params).await
+                    }
+                }
+                .await;
+
+                if let Err(err) = result {
+                    tracing::error!("proxy session error from {peer_address:?}: {err:#}");
+                }
+            });
+        }
+    }
+
+    async fn handle_client<S>(
+        stream: S,
+        peer_address: SocketAddr,
+        local_address: SocketAddr,
+        params: &ProxyListenerParams,
+    ) -> anyhow::Result<()>
+    where
+        S: AsyncRead + AsyncWrite + Unpin + Send + 'static,
+    {
+        crate::proxy_handler::handle_proxy_client(
+            stream,
+            peer_address,
+            local_address,
+            params.timeout,
+            params.no_splice,
+            params.require_auth,
+        )
+        .await
+    }
+}
+
+/// Connection metadata passed to the proxy_server_auth_rfc1929 callback
+#[derive(Clone, Debug, serde::Serialize)]
+pub struct ConnMeta {
+    pub peer_address: SocketAddr,
+    pub local_address: SocketAddr,
+}
+
+impl IntoLua for ConnMeta {
+    fn into_lua(self, lua: &Lua) -> mlua::Result<mlua::Value> {
+        lua.to_value(&self)
+    }
+}
+
+declare_event! {
+    static CHECK_AUTH: Single(
+        "proxy_server_auth_rfc1929",
+        username: String,
+        password: String,
+        conn_meta: ConnMeta,
+    ) -> SerdeWrappedValue<AuthKindResult>;
+}
+
+/// Validates credentials via the proxy_server_auth_rfc1929 Lua callback.
+/// The callback receives (username, password, conn_meta) where conn_meta is a table
+/// containing peer_address and local_address.
+/// Returns an AuthInfo on success (with the authenticated identity populated),
+/// or None if authentication failed.
+pub async fn authenticate_user(
```

Comment:
I'm normally in favor of breaking things out into separate functions like this, but I think that in this case it leads to more code than is necessary; I think that we'll have clearer and more concise code if this logic was inlined into the `perform_auth` function so that it could then directly consult the `AuthKindResult` and either communicate the authentication failure directly, or directly update its `auth_kind` state.